tests/Baseline/spicy.optimization.default-parser-functions/log

```diff
@@ -78,6 +78,12 @@
 [debug/optimizer] removing field for unused method foo::P2::__on_0x25_undelivered
 [debug/optimizer] removing field for unused method foo::P2::__on_x
 [debug/optimizer] removing field for unused method foo::P2::__str__
+[debug/optimizer] removing field for unused method foo::__register_foo_P1::<anon-struct-5>::parse1
+[debug/optimizer] removing field for unused method foo::__register_foo_P1::<anon-struct-5>::parse2
+[debug/optimizer] removing field for unused method foo::__register_foo_P1::<anon-struct-5>::parse3
+[debug/optimizer] removing field for unused method foo::__register_foo_P2::<anon-struct-8>::parse1
+[debug/optimizer] removing field for unused method foo::__register_foo_P2::<anon-struct-8>::parse2
+[debug/optimizer] removing field for unused method foo::__register_foo_P2::<anon-struct-8>::parse3
```

Comment:
I'm wondering what's happening here: these anon structs should appear only for struct constructors, but we'd be using their fields when we coerce it to the actual type. Would suggest to take a quick look at the before-/after-optimization code here to double-check what's changing.

Reply:
Yes, this one is weird, and it seemed surprising that this even works. The reason for this is in how struct ctrs are represented in the AST, i.e., code of this form:

```
foo::P1::__parser = [
  $name="foo::P1",
  $parse1=foo::P1::parse1,
  $parse2=foo::P1::parse2,
  $parse3=foo::P1::parse3,
  $context_new=Null, $type_info=typeinfo(P1),
  $description="",
  $mime_types=vector(),
  $ports=vector()
];
```

Simplified this looks like this:
```
- statement::Expression [@s:7fdad2171960]
  - expression::Assign (non-const) (resolved) [@e:7fdad2171650]
    - expression::ResolvedID <rid="250"> (type: spicy_rt::Parser [@t:7fdad043dd80]) (non-const) (resolved) [@e:7fdac05621b0]
      - ID <name="foo::P1::__parser"> [@i:7fdac0562140]
    - expression::Ctor (default-parser-functions.spicy:14:18) (const) (resolved) [@e:7fdad0616a40]
      - ctor::Coerced (default-parser-functions.spicy:14:18) [@c:7fdad0617200]
        - ctor::Struct (default-parser-functions.spicy:14:18) [@c:7fdad21a8570]
          - type::Struct %413 (default-parser-functions.spicy:14:18) (non-const) (resolved) [@t:7fdac0566530]
            - declaration::Expression %422 <linkage="private"> (default-parser-functions.spicy:14:18) [canon-id: foo::__register_foo_P1::<anon-struct-5>::self] [@d:7fdad0613e40]
              - ID <name="self"> (default-parser-functions.spicy:14:18) [@i:7fdad0613b20]
            - declaration::Field <cc="<standard>"> (default-parser-functions.spicy:14:18) [canon-id: foo::__register_foo_P1::<anon-struct-5>::name] [@d:7fdac0564850]
              - ID <name="name"> [@i:7fdac0564740]
            - declaration::Field <cc="<standard>"> (default-parser-functions.spicy:14:18) [canon-id: foo::__register_foo_P1::<anon-struct-5>::parse1] [@d:7fdac0564ce0]
              - ID <name="parse1"> [@i:7fdac0564900]
            [MORE FIELDS]

          - ctor::struct_::Field (default-parser-functions.spicy:14:18) [@c:7fdad21a9050]
            - ID <name="name"> [@i:7fdad2171ae0]
            - expression::Ctor (const) (resolved) [@e:7fdad216f3b0]
              - ctor::String <value="foo::P1"> [@c:7fdad216e250]
                - type::String (const) (resolved) [@t:7fdad216f570]
          - ctor::struct_::Field (default-parser-functions.spicy:14:18) [@c:7fdad21a9110]
            - ID <name="parse1"> [@i:7fdad2179ca0]
            - expression::ResolvedID <rid="314"> (type: function method view<stream> (inout value_ref<stream> data, optional<view<stream>> cur = Null, optional<spicy_rt::UnitContext> context) [@t:7fdad21725a0]) (const) (resolved) [@e:7fdac0563460]
              - ID <name="foo::P1::parse1"> [@i:7fdac05633f0]
            [MORE FIELDS]
```

We are removing some `declaration::Field`s which mirror data also present in the `ctor::struct_::Field`s so the data we need is still there in the end.

I am not really sure what to do here since this seems to actually remove dead code, but still surprising. Suggest we leave as is.